File: src/simulator.c
#include "simulator.h"
#include "dispatcher.h"
#include "process.h"

struct sim_struct {
    int     mode;
    dispatcher  dis;
    int     delay_mode;
    int     delay;
    int     cycles_since_pause;
    bool    io;
    bool    io_cycles;
    int     total_cycles;
}

simulator sim_init(int delay_mode, int delay, int quantum, char **filelist, int fl_len) {
    simulator s = malloc(sizeof(struct sim_struct));
    s->dis = dis_init(quantum, filelist, fl_len);
    s->sim_mode = KERNAL_MODE;
    s->delay_mode = delay_mode;
    s->delay = (delay > 0) ? delay : 1;
    s->io = false;
    s->total_cycles = 0;

    DEBUG_PRINT("[Simulator] Successfully initialized!");    
    return s;
}

void sim_run(simulator s) {
    s->total_cycles++;
    DEBUG_PRINT("[Simulator] Cycle %d:", s->total_cycles);

    if (!s->io && (rand() % 10) < 3) { // io generation
        DEBUG_PRINT("[Simulator] IO event generated.");
        s->io = true;
        s->io_cycles = (rand() % 25) + 25; // between 25 - 50
        dis_io(s->dis, true);
    }
    else if (s->io) { // set limit on how long io can last
        if (s->io_cycles <= 0) {
            s->io = false;
            dis_io(s->dis, false);
        }
        else
            s->io_cycles--;
    }

    switch (s->mode) {
        case KERNAL_MODE:
            DEBUG_PRINT("[Simulator] In Kernel Mode.");
            dis_schedule(s->dis); 
            s->mode = USER_MODE;
            break;
        case USER_MODE:
            DEBUG_PRINT("[Simulator] In User Mode.");
            s->mode = dis_run(s->dis); 
            break;
    }

    if (s->delay_mode == D_MODE_PAUSE) {
        if (s->cycles_since_pause == s->delay) {
            s->cycles_since_pause = 0;
            system("pause");
        }
        else
            s->cycles_since_pause++;
    }
    else if (delay mode == D_MODE_DELAY)
        sleep(s->delay);
}

File: src/simulator.py

from random import randint
from time import sleep

import src.commands as c

from src.globals import DebugPrint, sim_mailbox, gui_mailbox
from src.dispatcher import Dispatcher

def PauseExecution():
    print("Type exit to close or newline to continue")
    if input() == "exit":
        exit()

def Sim_MainThread():
    s = Simulator(10, list())
    while True:
        s.loop()

class Simulator:

    def __init__(self, quant, filelist):
        DebugPrint("[Simulator] Beginning initialization")

        self.dis = Dispatcher(quant, filelist)
        self.mode = False   # kernal mode
        self.cycles_since_pause = 0
        self.delay = 0
        self.io = False 
        self.io_cycles = 0
        self.total_cycles = 0
        self.on_flag = False
        self.refresh_time = 0.5
        self.clock_time = 0

        DebugPrint("[Simulator] Successfully initialized!")

    def checkMail(self):
        if sim_mailbox.empty():
            DebugPrint("[Simulator] No messages")
            return

        cmd = sim_mailbox.get()
        DebugPrint(f"[Simulator] Message: {cmd}")

        if cmd == "run":
            self.on_flag = True
        elif cmd == "stop":
            self.on_flag = False
        elif isinstance(cmd, c.Cmd_CreateProcess):
            for _ in range(cmd.num):
                self.dis.createProcess(cmd.filepath)
        elif isinstance(cmd, c.Cmd_Quantum):
            self.dis.quant = cmd.x
        elif cmd == "close":
            exit()
        elif isinstance(cmd, c.Cmd_ChangeSpeed):
            self.refresh_time = cmd.x
        elif cmd == "switch":
            self.dis.switch()
        elif isinstance(cmd, c.Cmd_RandProcess):
            self.dis.randomProcess(cmd.x)

    def loop(self):
        self.checkMail()
        if self.on_flag:
            self.run()
        sleep(self.refresh_time)
        self.clock_time += self.refresh_time

    def stats(self):
        return [f"Total cycles: {self.total_cycles}\n"] + self.dis.stats()

    def run(self):
        self.total_cycles += 1
        DebugPrint(f"[Simulator] Cycle {self.total_cycles}")
        
        if not self.io and randint(0, 99) < 5:
            self.io = True 
            self.io_cycles = randint(25, 50)
            self.dis.setIO(True)
            DebugPrint("[Simulator] IO event")
        elif self.io:
            if self.io_cycles <= 0:
                self.io = False
                self.dis.setIO(False)
            else:
                self.io_cycles -= 1

        if self.mode:   # user mode
            DebugPrint("[Simulator] In User Mode")
            self.mode = self.dis.runCurrent()
        else:           # kernal mode
            DebugPrint("[Simulator] In Kernel Mode")
            self.dis.schedule()
            self.mode = True

        if self.total_cycles % 5 == 0:
            gui_mailbox.put(c.Cmd_Stats(self.stats()))

                